rllib/algorithms/marwil/marwil_offline_prelearner.py

```diff
@@ -0,0 +1,214 @@
+import numpy as np
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+from ray.rllib.core.columns import Columns
+from ray.rllib.evaluation.postprocessing import Postprocessing
+from ray.rllib.offline.offline_data import OfflinePreLearner
+from ray.rllib.policy.sample_batch import MultiAgentBatch, SampleBatch
+from ray.rllib.utils.annotations import override, OverrideToImplementCustomLogic
+from ray.rllib.utils.numpy import convert_to_numpy
+from ray.rllib.utils.postprocessing.value_predictions import compute_value_targets
+from ray.rllib.utils.postprocessing.episodes import (
+    add_one_ts_to_episodes_and_truncate,
+    remove_last_ts_from_data,
+    remove_last_ts_from_episodes_and_restore_truncateds,
+)
+from ray.rllib.utils.postprocessing.zero_padding import unpad_data_if_necessary
+from ray.rllib.utils.typing import EpisodeType, TensorType
+
+
+class MARWILOfflinePreLearner(OfflinePreLearner):
+    @OverrideToImplementCustomLogic
+    @override(OfflinePreLearner)
+    def __call__(self, batch: Dict[str, np.ndarray]) -> Dict[str, MultiAgentBatch]:
+
+        # Map the batch to episodes.
+        episodes = self._map_to_episodes(self._is_multi_agent, batch, finalize=True)
+
+        # Compute the GAE via the Learner's method.
+        batch, episodes = self._compute_gae_from_episodes(episodes=episodes["episodes"])
+
+        # Run the `Learner`'s connector pipeline.
+        batch = self._learner_connector(
+            rl_module=self._module,
+            data=batch,
+            episodes=episodes,
+            shared_data={},
+        )
+
+        # Convert to `MultiAgentBatch`.
+        batch = MultiAgentBatch(
+            {
+                module_id: SampleBatch(module_data)
+                for module_id, module_data in batch.items()
+            },
+            # TODO (simon): This can be run once for the batch and the
+            # metrics, but we run it twice: here and later in the learner.
+            env_steps=sum(e.env_steps() for e in episodes),
+        )
+        # Remove all data from modules that should not be trained. We do
+        # not want to pass around more data than necessaty.
+        for module_id in list(batch.policy_batches.keys()):
+            if not self._should_module_be_updated(module_id, batch):
+                del batch.policy_batches[module_id]
+
+        # TODO (simon): Log steps trained for metrics (how?). At best in learner
+        # and not here. But we could precompute metrics here and pass it to the learner
+        # for logging. Like this we do not have to pass around episode lists.
+
+        # TODO (simon): episodes are only needed for logging here.
+        return {"batch": [batch]}
+
+    def _compute_gae_from_episodes(
```

Comment:
Any chance we can share this code with PPO's?

Create a utility function in PPO folder (e.g. algorithms/ppo/utils.py), then import and use this same function for MARWIL.

Reply:
That's the question I have: DO we really need the complete code? Or can we get rid of either the batch or episodes in their when considering MARWIL and the offline learning pipeline?